Display advanced block propagation technology (XThin/Graphene) statistics in the Debug Window UI

Justaphf opened the pull request:
Now that graphene is merged into `dev`, I wanted a quick way to monitor which block propagation technologies are being actively used directly in the QT UI.  This is just displaying the number of blocks sent/received by XThin/Graphene. Does not currently include any of the other more detailed information available via `getnetworkinfo`.  Stats are automatically updated in the UI using one of the pre-existing timers for the debug window.

Justaphf committed:
Hook up stat update to the UI

Justaphf committed:
Add labels to UI for block propagation stats

Justaphf committed:
Separately track inbound and outbound XThin blocks

This is so we can maintain and display the total send/receive count, in addition to the last 24-hour statistics (matching graphene).

sickpig reviewed (commented):
@Justaphf rebase needed (I know this is still WIP but still)

Justaphf committed:
Correct thinblock qa test

Test relies on exact output string format of 'thinblockstats'.'summary' field
which was changed in the previous commit

Justaphf committed:
Add quick stats struct to Graphene and helper methods to fill the data

This is to allow grabbing multiple statistics in one go for display in the UI while only taking a single lock on cs_graphenestats
The helper methods compute in exactly the same way the existing string methods do, and are intended to replace the internal computations to properly follow DRY methodology (subsequent commit).
NOTE: Set up to use clang static analysis LOCK macros.

Justaphf committed:
Update existing Graphene string methods to use new internal methods instead

This is such that any computations use a single algorithm

Justaphf committed:
Add quick stats struct to ThinBlock and helper methods to fill the data

This is to allow grabbing multiple statistics in one go for display in the UI while only taking a single lock on cs_thinblockstats
The helper methods compute in exactly the same way the existing string methods do, and are intended to replace the internal computations to properly follow DRY methodology (subsequent commit).
NOTE: Set up to use clang static analysis LOCK macros.

Justaphf committed:
Update existing ThinBlock string methods to use new internal methods instead

This is such that any computations use a single algorithm

Justaphf commented:
Taking out of WIP.  Everything is working now in terms of displaying what, IMO, were the most prominent summary statistics for both XThin and Graphene on the debug dialog.  This is a subset of the full statistics available via the `getnetworkinfo` RPC call.

Looking for feedback on the UI layout and category/label naming.  I tried to combine everything in a compact way that doesn't require the UI dialog size to be too large to see everything.  I grouped by technology type, then statistic time-frame, but maybe it makes more sense to group by time-frame then by technology type.  I tried it both ways and see benefits for both.

NOTE: As part of this, I updated the UI layout via Qt Creator.  All fields on the first tab are in a 3-column grid layout.  I changed it such that title rows (bold text) now span all 3 columns, label-value rows are defined as label 1 column, value spans 2 columns.  This caused a lot of churn in the debugwindow.ui file.  Not sure if making these changes caused some unintended side-effect with the UI layout, but seems to be OK on a Windows 10 test system.

Side-bar: I don't know that I like terming it "compression".  While technically it is a form of compression ("differential compression", on a highly ephemeral, per-destination specific source), it's not really compression in the sense that your average Joe, non-technical computer user thinks of (i.e. archives with zip files, image compression with jpeg/png/etc).  Technical users should understand the difference, but I think less technical users may be confused by the terminology.  Not sure how big of an issue this is (it's purely a UX thing), just wanted to note it since it came up as part of implementing this.

awemany commented on src/thinblock.cpp:
```diff
@@ -1072,8 +1072,8 @@ std::string CThinBlockData::ToString()
     std::ostringstream ss;
-    ss << nBlocks() << " thin " << ((nBlocks() > 1) ? "blocks have" : "block has") << " saved " << formatInfoUnit(size)
-       << " of bandwidth";
+    ss << nInBoundBlocks() << " inbound and " << nOutBoundBlocks() << " outbound thin blocks have saved "
```
Nit: Maybe `block(s)` instead of `block` to properly deal with the singular case :)

Justaphf replied:
I think that since there are always 2 numbers listed now it is grammatically correct to always use the plural.

awemany replied:
Indeed!

awemany commented on src/graphene.cpp:
```diff
@@ -792,6 +792,91 @@ double CGrapheneBlockData::average(std::map<int64_t, uint64_t> &map)
     return (double)accum / map.size();
 }
 
+double CGrapheneBlockData::computTotalBandwidthSavingsInternal() EXCLUSIVE_LOCKS_REQUIRED(cs_graphenestats)
```
Typo in the function name? 'comput'

Justaphf replied:
Will fix.  I also noticed some spelling/typos in some of the comments as well.

Justaphf replied:
Fixed.

awemany commented on src/graphene.cpp:
```diff
@@ -792,6 +792,91 @@ double CGrapheneBlockData::average(std::map<int64_t, uint64_t> &map)
+    double nCompressionRate = 0;
+    uint64_t nGrapheneSizeTotal = 0;
+    uint64_t nOriginalSizeTotal = 0;
+    for (std::map<int64_t, std::pair<uint64_t, uint64_t> >::iterator mi = mapGrapheneBlocksInBound.begin();
```
nit: IMO `for (auto|<type> m : ...) {...}` is more readable

Justaphf replied:
I hadn't changed the original algorithms beyond moving them into helper methods.   I can refactor this and any other locations where I touched the code.

Justaphf replied:
Fixed for all old style for loops in thinblock.cpp and graphene.cpp related to stats computations.

awemany replied:
Great!

awemany commented on src/graphene.cpp:
```diff
@@ -792,6 +792,91 @@ double CGrapheneBlockData::average(std::map<int64_t, uint64_t> &map)
+    return nCompressionRate;
+}
+
+double CGrapheneBlockData::compute24hOutboundCompressionInternal() EXCLUSIVE_LOCKS_REQUIRED(cs_graphenestats)
```
Can you further DRY that with the inbound function above? Seems like they share a lot of code still...

Justaphf replied:
Will do.

Justaphf replied:
Fixed

awemany commented on src/thinblock.cpp:
```diff
@@ -984,6 +984,93 @@ double CThinBlockData::average(std::map<int64_t, uint64_t> &map)
+    return nCompressionRate;
+}
+
+double CThinBlockData::compute24hOutboundCompressionInternal() EXCLUSIVE_LOCKS_REQUIRED(cs_thinblockstats)
```
Hmm, this is an almost repeat of `compute24hInboundCompressionInternal()`, isn't it?

Justaphf replied:
Will make the same DRY updates as requested for graphene.

Justaphf replied:
Fixed

awemany replied:
Great, other change as well!

awemany reviewed (commented):
Are the newly computed stats also available via RPC?

ptschip commented:
@Justaphf been running this today, the only thing i'd like to bikeshed about is the compression ratios having 2 decimal places rather than one in the getnetworkinfo.  I think the 2 just adds to visual noise and IMO we should be consistent with getnetworkinfo.  Other than that I think it all looks great!

Justaphf commented:
@ptschip I'm working on the comments from @awemany presently and will make the precision persistent for the compression percentages as well.

I also found a bug in the compression statistics computations for both XThin and Graphene that I'm fixing.  Looks like we expire the block maps at the time of the computation, but don't expire the bloom filter/mempool info maps, which are also used in the computation.  This could potentially cause extra values to remain in the totals for bloom filter/mempool info for a short window, producing slightly lower compression percentages.

Justaphf committed:
Update statistics map iterators to C++11 range based for loops

Justaphf committed:
Update all 24-hour average statistics to expirestats() on all maps used in computation

Some computations called expire stats on only one of the two maps which could potentially lead to the second map containing more data entries than first map, skewing the results.
Other computations did not expire stats at all on the map used in the computation which could potentially lead to the computed stat spanning slightly more than the expected 24-hour period.

Justaphf commented:
Fixed review comments.

Additionally corrected the stats computations to ensure that all stats maps used in computations have called `expireStats` prior to computation to ensure the data being computed correctly spans the specified 24-hour period.  Two potential errors were caught here.  Some of the computations use values from 2 maps, but only called `expireStats` on one of the maps.  Other computations didn't call `expireStats` at all so could exceed the 24-hour window.

NOTE: Several of the refactors I left as separate commits to make it easier to review.  Can squash up once approved for merge if so desired.

ptschip commented on src/thinblock.cpp:
```diff
@@ -1415,6 +1475,31 @@ void CThinBlockData::DeleteThinBlockBytes(uint64_t bytes, CNode *pfrom)
+
+    stats.fEnabled = IsThinBlocksEnabled();
+    if (!stats.fEnabled)
+        return;
```
why do we have to track if thinblocks is enabled by setting stats.fEnabled, when we can always get the current value from IsThinBlocksEnabled()?

Justaphf replied:
You are correct.  I will remove and directly use the Is*Enabled() calls for thin blocks and graphene in the UI display logic.

Justaphf commented:
Correct @ptschip last comment and rebased.

ptschip reviewed (approved):
looks good, thanks for doing this...
